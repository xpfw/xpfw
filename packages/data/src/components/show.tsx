import { LoadingStore } from "@xpfw/form-shared"
import { get } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore, { IResult } from "../store/db"

export interface ISharedFormShow extends React.Props<any> {
  collection?: string
  id?: string
}

export interface ISharedFormShowState {
  item?: IResult
  loading: boolean
}

export interface IFormShowProps extends ISharedFormShow, ISharedFormShowState {
}

function SharedFormShow<T>(Component: React.ComponentType<IFormShowProps & T>):
React.ComponentType<ISharedFormShow & T> {
  return class extends ComponentBase<ISharedFormShow & T, ISharedFormShowState> {
    public render() {
      return <Component {...this.props} {...this.state} />
    }
    protected _buildState(props: ISharedFormShow, initialBuild: boolean): ISharedFormShowState {
      const id: any = get(props, "id", "none")
      const collection: any = get(props, "collection", "none")
      const loading = LoadingStore.getLoading(id)
      return {
        item: get(DbStore.getGetState(id, collection, true), "result"),
        loading
      }
    }
  }
}

export default SharedFormShow
