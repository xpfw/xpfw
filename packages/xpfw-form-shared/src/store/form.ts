import {
  globals, IField, IFieldError, IForm, IFormError, iterateFields,
  validateField, validateForm, ValidationRegistry
} from "@xpfw/validate"
import { get, isFunction, isNil, isObject, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"
import FormErrorStore from "./error"

@AutoSubscribeStore
export class FormStore extends StoreBase implements IPersistableStore {
  public name = "formData"
  private formData: any = {}

  public getPropKeys() { return ["formData"] }

  public setValue(valuePath: string, value: any, fullTrigger: boolean = false) {
    set(this.formData, valuePath, value)
    this.trigger(fullTrigger ? undefined : valuePath)
  }

  @autoSubscribe
  public getValue(@key valuePath: string): any {
    return get(this.formData, valuePath)
  }

  public getFormData(form: IForm, prefix: string = "", method: string = "create") {
    const fullData: any = {}
    const formData = this.formData
    iterateFields(form, (field) => {
      let defaultValue = get(field, "validate.defaultValue")
      if (method === globals.Method.Find) {
        defaultValue = undefined
      }
      let value = get(formData,
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, defaultValue)
      let converter: any = get(field, "validate.convert")
      if (!isFunction(converter) && isObject(converter)) {
        converter = converter[method]
      }
      if (isFunction(converter)) {
        value = converter(value)
      }
      if (method !== globals.Method.Find || method === globals.Method.Find && !isNil(value))  {
        set(fullData, field.mapTo, value)
      }
    })
    return fullData
  }

  public copyDocToFormData(doc: any, form: IForm, prefix: string = "") {
    iterateFields(form, (field) => {
        this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, get(doc, field.mapTo))
    })
    this.trigger()
  }

  public async validateField(field: IField, form: IForm, method: string = "create",
                             prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`
    try {
      await validateField(get(this.formData, valuePath), field, {
        method,
        field,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public async validateForm(form: IForm, method: string = "create", prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${form.model}`
    try {
      await validateForm(this.getFormData(form, prefix), form, {
        method,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public resetForm(form: IForm, prefix: string = "")  {
    iterateFields(form, (field) => {
      this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, undefined)
    })
  }
}

export default new FormStore()
