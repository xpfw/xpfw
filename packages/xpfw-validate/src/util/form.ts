import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
