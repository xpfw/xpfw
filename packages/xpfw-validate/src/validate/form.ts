import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
