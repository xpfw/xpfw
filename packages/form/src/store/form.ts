import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
