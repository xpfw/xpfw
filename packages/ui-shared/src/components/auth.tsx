import { FormErrorStore, FormStore } from "@xpfw/form-shared"
import { globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore from "../store/db"
import UserStore from "../store/user"

const FieldType = globals.FieldType
const RequiredType = globals.RequiredType

const MailField: IField = {
  mapTo: "email",
  type: FieldType.Text,
  validate: {required: {type: RequiredType.Always}}
}
const PwField: IField = {
  mapTo: "password",
  type: FieldType.Password,
  validate: {required: {type: RequiredType.Always}}
}

/**
 * Be sure to adjust relationship(Id|NamePath) to your user object
 */
const OwnerField: IField = {
  mapTo: "belongsTo",
  type: FieldType.RelationshipSingle,
  validate: {
    relationshipNamePath: MailField.mapTo,
    relationshipCollection: "user"
  }
}

const AuthForm: IForm = {
  model: "authForm",
  collection: "user",
  sections: [{fields: [MailField, PwField]}]
}

export interface ISharedFormAuth extends React.Props<any> {
}

export interface ISharedFormAuthState {
  error?: any
  user?: any
  connected: boolean
  loggedIn: boolean
  loading: boolean
}

export interface IFormAuthProps extends ISharedFormAuth, ISharedFormAuthState {
  submitLogin: any
  submitRegister: any
  submitLogout: any
}
const submitLogin = () => UserStore.login()
const submitRegister = async () => {
  const registerRes = await UserStore.register()
  await submitLogin()
  return registerRes
}
const submitLogout = () => UserStore.logout()

function SharedFormAuth<T>(Component: React.ComponentType<IFormAuthProps & T>):
  React.ComponentType<ISharedFormAuth & T> {
  return class extends ComponentBase<ISharedFormAuth & T, ISharedFormAuthState> {
    public render() {
      const customProps = {
        submitLogin, submitRegister, submitLogout
      }
      return <Component {...this.props} {...this.state} {...customProps} />
    }
    protected _buildState(props: ISharedFormAuth, initialBuild: boolean): ISharedFormAuthState {
      return {
        loading: UserStore.getLoading(),
        user: UserStore.getUser(),
        loggedIn: UserStore.getLoggedIn(),
        error: UserStore.getAuthErr(),
        connected: UserStore.getConnected()
      }
    }
  }
}

export default SharedFormAuth
export {
  AuthForm, PwField, MailField, OwnerField,
  submitLogin, submitRegister, submitLogout
}
