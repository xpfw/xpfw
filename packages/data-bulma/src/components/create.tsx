import * as React from "react"

import { dataOptions,  ICreateHookProps, useCreateWithProps } from "@xpfw/data"
import { getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"

const BulmaCreate: React.FunctionComponent<ICreateHookProps> = observer((props) => {
  const createProps = useCreateWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = createProps.error != null
  const result = createProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <div className="notification is-danger">
        Error please recheck your inputs or connection {JSON.stringify(createProps.error)}
      </div>
    )
  }
  if (result) {
    msg = (
      <div className="notification is-success">
        Successfully created {get(result, dataOptions.idPath)}
      </div>
    )
  }
  return (
    <div>
      {fields}
      <a className="button is-primary" onClick={createProps.submitCreate}>Create</a>
      {msg}
    </div>
  )

})

export default BulmaCreate
export {
  ICreateHookProps
}
