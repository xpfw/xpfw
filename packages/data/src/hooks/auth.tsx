import { ExtendedJSONSchema, JSONSchemaDefinition } from "@xpfw/form"
import UserStore from "../store/user"

const MailField: JSONSchemaDefinition = {
  title: "email",
  type: "string"
}
const PwField: JSONSchemaDefinition = {
  title: "password",
  type: "string",
  format: "password"
}

/**
 * Be sure to adjust relationship(Id|NamePath) to your user object
 */
const OwnerField: ExtendedJSONSchema = {
  title: "belongsTo",
  type: "relationshipSingle",
  relationship: {
    namePath: MailField.title,
    collection: "user"
  }
}

const AuthForm: ExtendedJSONSchema = {
  title: "authForm",
  collection: "user",
  required: [String(MailField.title), String(PwField.title)],
  properties: {
    [String(MailField.title)]: MailField,
    [String(PwField.title)]: PwField
  }
}

const submitLogin = () => UserStore.login()
const submitRegister = async () => {
  const registerRes = await UserStore.register()
  await submitLogin()
  return registerRes
}
const submitLogout = () => UserStore.logout()

const useAuth = () => {
  return {
    loading: UserStore.getLoading(),
    user: UserStore.getUser(),
    loggedIn: UserStore.getLoggedIn(),
    error: UserStore.getAuthErr(),
    connected: UserStore.getConnected(),
    submitLogout, submitLogin, submitRegister
  }
}

export default useAuth
export {
  AuthForm, PwField, MailField, OwnerField,
  submitLogin, submitRegister, submitLogout
}
