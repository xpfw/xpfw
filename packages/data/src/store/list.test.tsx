import { FeathersClient } from "@xpfw/data-feathers"
import { toJS } from "@xpfw/data-tests"
import { FormStore } from "@xpfw/form"
import {
  makeSubFields, NameField, NumberAndRequiredTextSchema, NumberField,
  RelationshipAndNumberSchema, RelationshipMultiField, RelationshipSingleField
 } from "@xpfw/form-tests"
import { getRandomApp } from "@xpfw/test-util"
import "isomorphic-fetch"
import { cloneDeep, get, isNil } from "lodash"
import BackendClient from "../client"
import { AuthForm, MailField, PwField } from "../components/auth"
import DbStore from "./db"
import ListStore from "./list"

BackendClient.client = FeathersClient

test("List filter Test", async () => {
  const cols: any = ["simpleTestCol", RelationshipAndNumberSchema.collection]
  const appRef = await getRandomApp(cols, false, BackendClient.client, false)
  const fields = makeSubFields(NumberAndRequiredTextSchema)
  const relFields = makeSubFields(RelationshipAndNumberSchema)
  for (let i = 0; i < 450; i++) {
    fields[NameField.title].setValue(`${i}*${i}`)
    fields[NumberField.title].setValue(i)
    const res: any = await DbStore.create(NumberAndRequiredTextSchema)
    relFields[RelationshipSingleField.title].setValue(res.id)
    relFields[NumberField.title].setValue(i * 420)
    await DbStore.create(RelationshipAndNumberSchema)
  }
  FormStore.setValue(NumberAndRequiredTextSchema.title, undefined)
  FormStore.setValue(RelationshipAndNumberSchema.title, undefined)
  expect(await ListStore.makeQuery(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "page"))
    .toMatchSnapshot("result of initial search")
  expect(await ListStore.searchInPage(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "page", 99999))
    .toMatchSnapshot("result of searching too far")
  expect(await ListStore.searchInPage(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "page", 4))
    .toMatchSnapshot("result of searching in higher page")
  let findFields = makeSubFields(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "find")
  findFields[NameField.title].setValue("4*4")
  expect(await ListStore.makeQuery(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "find"))
    .toMatchSnapshot("result of search with text")
  const find2Fields = makeSubFields(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "asdf")
  find2Fields[NumberField.title].setValue(240)
  expect(await ListStore.makeQuery(NumberAndRequiredTextSchema, NumberAndRequiredTextSchema.title, "asdf"))
    .toMatchSnapshot("result of search with number")

  const multiFormStore = cloneDeep(NumberAndRequiredTextSchema)
  delete multiFormStore.collection
  multiFormStore.properties[String(RelationshipSingleField.title)] = RelationshipSingleField
  multiFormStore.multiCollection = cols
FormStore.setValue(multiFormStore.title, undefined, "find")
  findFields = makeSubFields(multiFormStore, multiFormStore.title, "find")
  expect(await ListStore.makeQuery(multiFormStore, undefined, "page"))
    .toMatchSnapshot("result of initial search")
  expect(await ListStore.searchInPage(multiFormStore, undefined, "page", 999))
    .toMatchSnapshot("result of searching too far")
  expect(await ListStore.searchInPage(multiFormStore, undefined, "page", 4))
    .toMatchSnapshot("result of searching in higher page")
  findFields[NameField.title].setValue("6x6")
  expect(await ListStore.makeQuery(multiFormStore, undefined,  "find"))
    .toMatchSnapshot("result of search with text")
  findFields[multiFormStore.title].setValue(undefined)
  findFields[RelationshipSingleField.title].setValue(6)
  expect(await ListStore.makeQuery(multiFormStore, undefined, "find"))
    .toMatchSnapshot("result of search with ownedBy")
  findFields[multiFormStore.title].setValue(undefined)
  findFields[NumberField.title].setValue(420)
  expect(await ListStore.makeQuery(multiFormStore, undefined, "find"))
    .toMatchSnapshot("result of search with specifc number")
  // expect(toJS(FormStore)).toMatchSnapshot("form store after all query data filled in")
  await appRef.cleanUp()
}, 10000)
