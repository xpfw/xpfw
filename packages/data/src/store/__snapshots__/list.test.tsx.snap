// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List filter Test: result of initial search 1`] = `
Object {
  "data": Array [
    Object {
      "guestName": "0*0",
      "id": 0,
      "myVal": 0,
    },
    Object {
      "guestName": "1*1",
      "id": 1,
      "myVal": 1,
    },
    Object {
      "guestName": "2*2",
      "id": 2,
      "myVal": 2,
    },
    Object {
      "guestName": "3*3",
      "id": 3,
      "myVal": 3,
    },
    Object {
      "guestName": "4*4",
      "id": 4,
      "myVal": 4,
    },
    Object {
      "guestName": "5*5",
      "id": 5,
      "myVal": 5,
    },
    Object {
      "guestName": "6*6",
      "id": 6,
      "myVal": 6,
    },
    Object {
      "guestName": "7*7",
      "id": 7,
      "myVal": 7,
    },
    Object {
      "guestName": "8*8",
      "id": 8,
      "myVal": 8,
    },
    Object {
      "guestName": "9*9",
      "id": 9,
      "myVal": 9,
    },
  ],
  "limit": 10,
  "skip": 0,
  "total": 450,
}
`;

exports[`List filter Test: result of initial search 2`] = `
Object {
  "data": Array [
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "40*40",
      "id": 40,
      "myVal": 40,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "41*41",
      "id": 41,
      "myVal": 41,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "42*42",
      "id": 42,
      "myVal": 42,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "43*43",
      "id": 43,
      "myVal": 43,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "44*44",
      "id": 44,
      "myVal": 44,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "45*45",
      "id": 45,
      "myVal": 45,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "46*46",
      "id": 46,
      "myVal": 46,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "47*47",
      "id": 47,
      "myVal": 47,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "48*48",
      "id": 48,
      "myVal": 48,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "49*49",
      "id": 49,
      "myVal": 49,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 40,
      "myVal": 16800,
      "ownedBy": 40,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 41,
      "myVal": 17220,
      "ownedBy": 41,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 42,
      "myVal": 17640,
      "ownedBy": 42,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 43,
      "myVal": 18060,
      "ownedBy": 43,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 44,
      "myVal": 18480,
      "ownedBy": 44,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 45,
      "myVal": 18900,
      "ownedBy": 45,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 46,
      "myVal": 19320,
      "ownedBy": 46,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 47,
      "myVal": 19740,
      "ownedBy": 47,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 48,
      "myVal": 20160,
      "ownedBy": 48,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 49,
      "myVal": 20580,
      "ownedBy": 49,
    },
  ],
  "limit": 10,
  "skip": 40,
  "total": 450,
}
`;

exports[`List filter Test: result of search with number 1`] = `
Object {
  "data": Array [
    Object {
      "guestName": "240*240",
      "id": 240,
      "myVal": 240,
    },
  ],
  "limit": 10,
  "skip": 0,
  "total": 1,
}
`;

exports[`List filter Test: result of search with ownedBy 1`] = `
Object {
  "data": Array [
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 6,
      "myVal": 2520,
      "ownedBy": 6,
    },
  ],
  "limit": 10,
  "skip": 0,
  "total": 1,
}
`;

exports[`List filter Test: result of search with specifc number 1`] = `
Object {
  "data": Array [
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "420*420",
      "id": 420,
      "myVal": 420,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 1,
      "myVal": 420,
      "ownedBy": 1,
    },
  ],
  "limit": 10,
  "skip": 0,
  "total": 1,
}
`;

exports[`List filter Test: result of search with text 1`] = `
Object {
  "data": Array [
    Object {
      "guestName": "4*4",
      "id": 4,
      "myVal": 4,
    },
  ],
  "limit": 10,
  "skip": 0,
  "total": 1,
}
`;

exports[`List filter Test: result of search with text 2`] = `
Object {
  "data": Array [],
  "limit": 10,
  "skip": 0,
  "total": 0,
}
`;

exports[`List filter Test: result of searching in higher page 1`] = `
Object {
  "data": Array [
    Object {
      "guestName": "40*40",
      "id": 40,
      "myVal": 40,
    },
    Object {
      "guestName": "41*41",
      "id": 41,
      "myVal": 41,
    },
    Object {
      "guestName": "42*42",
      "id": 42,
      "myVal": 42,
    },
    Object {
      "guestName": "43*43",
      "id": 43,
      "myVal": 43,
    },
    Object {
      "guestName": "44*44",
      "id": 44,
      "myVal": 44,
    },
    Object {
      "guestName": "45*45",
      "id": 45,
      "myVal": 45,
    },
    Object {
      "guestName": "46*46",
      "id": 46,
      "myVal": 46,
    },
    Object {
      "guestName": "47*47",
      "id": 47,
      "myVal": 47,
    },
    Object {
      "guestName": "48*48",
      "id": 48,
      "myVal": 48,
    },
    Object {
      "guestName": "49*49",
      "id": 49,
      "myVal": 49,
    },
  ],
  "limit": 10,
  "skip": 40,
  "total": 450,
}
`;

exports[`List filter Test: result of searching in higher page 2`] = `
Object {
  "data": Array [
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "40*40",
      "id": 40,
      "myVal": 40,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "41*41",
      "id": 41,
      "myVal": 41,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "42*42",
      "id": 42,
      "myVal": 42,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "43*43",
      "id": 43,
      "myVal": 43,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "44*44",
      "id": 44,
      "myVal": 44,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "45*45",
      "id": 45,
      "myVal": 45,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "46*46",
      "id": 46,
      "myVal": 46,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "47*47",
      "id": 47,
      "myVal": 47,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "48*48",
      "id": 48,
      "myVal": 48,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "49*49",
      "id": 49,
      "myVal": 49,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 40,
      "myVal": 16800,
      "ownedBy": 40,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 41,
      "myVal": 17220,
      "ownedBy": 41,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 42,
      "myVal": 17640,
      "ownedBy": 42,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 43,
      "myVal": 18060,
      "ownedBy": 43,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 44,
      "myVal": 18480,
      "ownedBy": 44,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 45,
      "myVal": 18900,
      "ownedBy": 45,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 46,
      "myVal": 19320,
      "ownedBy": 46,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 47,
      "myVal": 19740,
      "ownedBy": 47,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 48,
      "myVal": 20160,
      "ownedBy": 48,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 49,
      "myVal": 20580,
      "ownedBy": 49,
    },
  ],
  "limit": 10,
  "skip": 40,
  "total": 450,
}
`;

exports[`List filter Test: result of searching too far 1`] = `
Object {
  "data": Array [
    Object {
      "guestName": "440*440",
      "id": 440,
      "myVal": 440,
    },
    Object {
      "guestName": "441*441",
      "id": 441,
      "myVal": 441,
    },
    Object {
      "guestName": "442*442",
      "id": 442,
      "myVal": 442,
    },
    Object {
      "guestName": "443*443",
      "id": 443,
      "myVal": 443,
    },
    Object {
      "guestName": "444*444",
      "id": 444,
      "myVal": 444,
    },
    Object {
      "guestName": "445*445",
      "id": 445,
      "myVal": 445,
    },
    Object {
      "guestName": "446*446",
      "id": 446,
      "myVal": 446,
    },
    Object {
      "guestName": "447*447",
      "id": 447,
      "myVal": 447,
    },
    Object {
      "guestName": "448*448",
      "id": 448,
      "myVal": 448,
    },
    Object {
      "guestName": "449*449",
      "id": 449,
      "myVal": 449,
    },
  ],
  "limit": 10,
  "skip": 440,
  "total": 450,
}
`;

exports[`List filter Test: result of searching too far 2`] = `
Object {
  "data": Array [
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "440*440",
      "id": 440,
      "myVal": 440,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "441*441",
      "id": 441,
      "myVal": 441,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "442*442",
      "id": 442,
      "myVal": 442,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "443*443",
      "id": 443,
      "myVal": 443,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "444*444",
      "id": 444,
      "myVal": 444,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "445*445",
      "id": 445,
      "myVal": 445,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "446*446",
      "id": 446,
      "myVal": 446,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "447*447",
      "id": 447,
      "myVal": 447,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "448*448",
      "id": 448,
      "myVal": 448,
    },
    Object {
      "fromCollection": "simpleTestCol",
      "guestName": "449*449",
      "id": 449,
      "myVal": 449,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 440,
      "myVal": 184800,
      "ownedBy": 440,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 441,
      "myVal": 185220,
      "ownedBy": 441,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 442,
      "myVal": 185640,
      "ownedBy": 442,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 443,
      "myVal": 186060,
      "ownedBy": 443,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 444,
      "myVal": 186480,
      "ownedBy": 444,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 445,
      "myVal": 186900,
      "ownedBy": 445,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 446,
      "myVal": 187320,
      "ownedBy": 446,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 447,
      "myVal": 187740,
      "ownedBy": 447,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 448,
      "myVal": 188160,
      "ownedBy": 448,
    },
    Object {
      "fromCollection": "myRelationshipCollection",
      "id": 449,
      "myVal": 188580,
      "ownedBy": 449,
    },
  ],
  "limit": 10,
  "skip": 440,
  "total": 450,
}
`;
